jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -452,26 +453,27 @@ static List<Locale> getLocales(Request request)
     {
         HttpFields fields = request.getHeaders();
         if (fields == null)
-            return List.of(Locale.getDefault());
+            return DEFAULT_LOCALES;
 
         List<String> acceptable = fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE);
 
-        // handle no locale
-        if (acceptable.isEmpty())
-            return List.of(Locale.getDefault());
-
-        return acceptable.stream().map(language ->
+        // return sorted list of locals, with known locales in quality order before unknown locales in quality order
+        return switch (acceptable.size())
         {
-            language = HttpField.stripParameters(language);
-            String country = "";
-            int dash = language.indexOf('-');
-            if (dash > -1)
+            case 0 -> DEFAULT_LOCALES;
+            case 1 -> List.of(Locale.forLanguageTag(acceptable.get(0)));
```

Comment:
We cannot call `Locale.forLanguageTag` for any string coming from an end-user as that's a potential CVE.

`Locale.forLanguageTag` will accept any 2+ letters alphabetical string and internally create and permanently store a `Locale` instance, without any limit.

The following test will OOME without any way to recover:
```java
    @Test
    public void testOomeLocale()
    {
        int index = 26; // start at 'AA'
        for (int i = 0; i < Integer.MAX_VALUE; i++)
        {
            String tag = toAlpha(index + i);
            Locale.forLanguageTag(tag);
        }
    }

    private static String toAlpha(int cp)
    {
        int div = cp / 26;
        int remainder = cp % 26;
        char letter = (char)((int)'A' + remainder);

        if (div != 0)
            return toAlpha(div - 1) + letter;
        return Character.toString(letter);
    }
```

Reply:
OH!  yes! Good spot, as the javadoc is not clear that caching is happening.    

So perhaps we go back to the impl I had where we just look up in a Trie index.   But we lose the ability to do deprecated name conversion then.
Hmmm must be a non allocating way to access that logic?  standby